File: vtkSPS/IO/General/Testing/Cxx/TestPolyDataPieceFilter.cxx
#include <spsPolyDataPieceFilter.h>
#include <vtkAppendPolyData.h>
#include <vtkCleanPolyData.h>
#include <vtkInformation.h>
#include <vtkNew.h>
#include <vtkSphereSource.h>
#include <vtkStreamingDemandDrivenPipeline.h>
#include <vtkTestUtilities.h>
#include <vtkTesting.h>

namespace
{
//------------------------------------------------------------------------------
void TestExpectedOutput(spsPolyDataPieceFilter* pieceFilter, const vtkIdType& nPieces,
  vtkIdType& numberOfPoints, vtkIdType& numberOfCells, const double tolerance = 1e-6)
{
  vtkInformation* outInfo = pieceFilter->GetExecutive()->GetOutputInformation(0);
  outInfo->Set(vtkStreamingDemandDrivenPipeline::UPDATE_NUMBER_OF_PIECES(), nPieces);

  vtkNew<vtkAppendPolyData> appendFilter;

  for (int iPiece = 0; iPiece < nPieces; iPiece++)
  {
    outInfo->Set(vtkStreamingDemandDrivenPipeline::UPDATE_PIECE_NUMBER(), iPiece);
    int upstreamPieceNumber =
      pieceFilter->GetInputAlgorithm(0, 0)->GetOutputPortInformation(0)->Get(
        vtkStreamingDemandDrivenPipeline::UPDATE_PIECE_NUMBER());
    std::cout << "UPDATE_PIECE_NUMBER(spsPolyDataPieceFilter): " << iPiece << std::endl;
    std::cout << "UPDATE_PIECE_NUMBER(vtkSphereSource): " << upstreamPieceNumber << std::endl;
    pieceFilter->Update();
    vtkNew<vtkPolyData> polyData;
    polyData->DeepCopy(pieceFilter->GetOutput());
    appendFilter->AddInputData(polyData);
  }
  vtkNew<vtkCleanPolyData> cleanFilter;
  cleanFilter->SetInputConnection(appendFilter->GetOutputPort());
  // We allow some tolerance
  cleanFilter->SetTolerance(1e-6);

  vtkPolyDataAlgorithm* polyDataAlgorithm = nullptr;
  // Uncomment for skipping cleaning of polydata
  // polyDataAlgorithm = vtkPolyDataAlgorithm::SafeDownCast(appendFilter);
  polyDataAlgorithm = vtkPolyDataAlgorithm::SafeDownCast(cleanFilter);
  polyDataAlgorithm->Update();
  numberOfPoints = polyDataAlgorithm->GetOutput()->GetNumberOfPoints();
  numberOfCells = polyDataAlgorithm->GetOutput()->GetNumberOfCells();
  std::cout << "Number of points: " << numberOfPoints << std::endl;
  std::cout << "Number of cells: " << numberOfCells << std::endl;
}
}

//------------------------------------------------------------------------------
int TestPolyDataPieceFilter(int argc, char* argv[])
{
  vtkIdType nPieces = 4;
  vtkNew<vtkSphereSource> sphereSource;
  sphereSource->SetThetaResolution(8);
  sphereSource->SetPhiResolution(8);
  sphereSource->Update();

  vtkNew<spsPolyDataPieceFilter> pieceFilter;

  vtkIdType nExpectedPoints = sphereSource->GetOutput()->GetNumberOfPoints();
  vtkIdType nExpectedCells = sphereSource->GetOutput()->GetNumberOfCells();
  vtkIdType nPoints = 0;
  vtkIdType nCells = 0;

  int retval = EXIT_SUCCESS;

  // Test trivial producer - we handle subdivision, so no tolerance
  pieceFilter->SetInputData(sphereSource->GetOutput());
  pieceFilter->UpdateInformation();
  TestExpectedOutput(pieceFilter, nPieces, nPoints, nCells, 0.0);
  retval |= !((nExpectedCells == nCells) && (nExpectedPoints == nPoints));

  // Test streaming producer capable of handling piece request
  pieceFilter->SetInputConnection(sphereSource->GetOutputPort());
  pieceFilter->UpdateInformation();
  TestExpectedOutput(pieceFilter, nPieces, nPoints, nCells);
  retval |= !((nExpectedCells == nCells) && (nExpectedPoints == nPoints));

  return retval;
}

File: vtkSPS/IO/General/Testing/Python/TestPolyDataPieceFilter.py
import addpaths

from spsmodules.spsIOGeneral import spsPolyDataPieceFilter
from spsmodules.util.scene_utils import vtk_color_all_cells

# noinspection PyUnresolvedReferences
import vtkmodules.vtkInteractionStyle
# noinspection PyUnresolvedReferences
import vtkmodules.vtkRenderingOpenGL2

from vtkmodules.vtkCommonDataModel import vtkPolyData
from vtkmodules.vtkCommonCore import vtkUnsignedCharArray
from vtkmodules.vtkCommonExecutionModel import vtkStreamingDemandDrivenPipeline
from vtkmodules.vtkFiltersCore import vtkAppendPolyData
from vtkmodules.vtkFiltersGeometry import vtkGeometryFilter
from vtkmodules.vtkFiltersSources import vtkSphereSource
from vtkmodules.vtkRenderingCore import (
    vtkRenderWindow,
    vtkActor,
    vtkRenderWindowInteractor,
    vtkRenderer,
    vtkDataSetMapper,
    vtkPolyDataMapper)
from vtkmodules.vtkFiltersParallel import vtkPieceRequestFilter

import numpy as np
import vtk

sphereSource = vtkSphereSource();
sphereSource.SetThetaResolution(30);
sphereSource.SetPhiResolution(30);
sphereSource.Update();

pieceFilter = spsPolyDataPieceFilter();

pieceFilter.SetInputData(sphereSource.GetOutput()) # We split into pieces
#pieceFilter.SetInputConnection(sphereSource.GetOutputPort()) # We split into pieces

# Renderer and window setup
renderer = vtkRenderer()
renderWindow = vtkRenderWindow()
renderWindow.AddRenderer(renderer)
interactor = vtkRenderWindowInteractor()
interactor.SetRenderWindow(renderWindow)

pieceFilter.UpdateInformation()

# Set up the number of pieces (streaming)
nPieces = 16
outInfo = pieceFilter.GetExecutive().GetOutputInformation(0)
outInfo.Set(vtkStreamingDemandDrivenPipeline.UPDATE_NUMBER_OF_PIECES(), nPieces)

append = vtkAppendPolyData()
append.SetInputConnection(pieceFilter.GetOutputPort())
for iPiece in range(nPieces):
    outInfo.Set(vtkStreamingDemandDrivenPipeline.UPDATE_PIECE_NUMBER(), iPiece)
    pieceFilter.Update()
    print("UPDATE_PIECE_NUMBER(spsPolyDataPieceFilter): %d" % (iPiece))
    iUpstreamPiece = sphereSource.GetOutputInformation(0).Get(vtkStreamingDemandDrivenPipeline.UPDATE_PIECE_NUMBER())
    print("UPDATE_PIECE_NUMBER(vtkSphereSource): " + str(iUpstreamPiece))
    # Take a copy
    polyData = vtkPolyData()
    polyData.DeepCopy(pieceFilter.GetOutput())
    #append.AddInputData(polyData)
    append.Update()
    # Debug
    print("#cells: " + str(polyData.GetNumberOfCells()))
    print("#points: " + str(polyData.GetNumberOfPoints()))
    # Color the mesh
    vtk_color_all_cells(polyData)
    # Display
    mapper = vtkPolyDataMapper()
    mapper.SetInputData(polyData)
    mapper.SetScalarModeToUseCellData()
    actor = vtkActor()
    actor.SetMapper(mapper)
    #renderer.AddActor(actor)
actor = vtkActor()


# Could use append filter and geometry filter (no deepcopy). Leave only outer shell as poly data
mapper = vtkDataSetMapper()
mapper.SetInputData(append.GetOutput())
actor.SetMapper(mapper)
renderer.AddActor(actor)

# Start rendering
renderWindow.Render()
interactor.Start()

